tests/integration/hsm-integration-test.cpp

```diff
@@ -0,0 +1,452 @@
+         * - Hash function's digest size doesn't match the message's digest size
+         * - Invalid signature format set
+         * - Signature can't be parsed
+         * - Invalid signature
+         */
+        std::cerr << "Verification failed!" << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+
+int main(void)
+{
+    // Information for engine loading and key management.
+    std::string id("pkcs11");
+    std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
+    std::string tokenLabel("token-label");
+    // Don't hardcode the pin in your application, this is just for demonstration purposes
+    std::string pin("1234");
+    HsmEngine hsmEngine(id, modulePath, tokenLabel, pin);
+    std::vector<uint8_t> message = utility::fromHex("deadbeef");
+
+    /************** ECC key generation and ECDSA **************/
+    std::vector<uint8_t> keyIDECC{};
+    std::string keyLabelECC("ecc-key-label");
+    auto ecdsaDigestType = DigestTypes::SHA512;
+    ECCSpec ecspec;
+    auto eccPrivKey =
+            AsymmetricPrivateKey::generateKeyOnHSM(hsmEngine, ecspec, keyLabelECC, keyIDECC);
+    auto ecdsaSigFormat = ECDSASignatureFormat::ASN1_SEQUENCE_OF_INTS;
+
+    /* The argument hashFunction is optional. Default is SHA256
+     * The default signature format is ECDSASignatureFormat::ASN1_SEQUENCE_OF_INTS */
+    auto ECDSAsignature = ecdsaSign(eccPrivKey, ecdsaDigestType, ecdsaSigFormat, message);
+
+    /* Use-Case 1: You want to check your own signature.
+     * We can use here the private key, as it also contains the public key.
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit upcast here.
+     */
+    ecdsaVerify(eccPrivKey, ecdsaDigestType, ecdsaSigFormat, ECDSAsignature, message);
+
+    /* Use-Case 2: The public key used for verification is stored in the HSM */
+    auto pubKeyEcc = AsymmetricPublicKey::readPublicKeyFromHSM(hsmEngine, keyLabelECC, keyIDECC);
+    ecdsaVerify(pubKeyEcc, ecdsaDigestType, ecdsaSigFormat, ECDSAsignature, message);
+
+    /* Use-Case 3: You want to write the public key to a PEM file and use it later
+     * for verification
+     */
+    auto pubKeyPem = eccPrivKey.publicKeyToPem();
+    /* Write PEM data to a file and read it again */
+    auto pubKeyEccFromPem = mococrw::AsymmetricPublicKey::readPublicKeyFromPEM(pubKeyPem);
+    ecdsaVerify(pubKeyEccFromPem, ecdsaDigestType, ecdsaSigFormat, ECDSAsignature, message);
+    /*********************************************/
+
+    /************** RSA key generation, loading and digital signatures **************/
+    std::vector<uint8_t> keyIDRSA{0x12, 0x34};
+    std::string keyLabelRSA{"rsa-key-label"};
+    auto mgf1DigestType = DigestTypes::SHA256;
+    mococrw::RSASpec rsaSpec;
+    auto rsaSignatureDigestType = DigestTypes::SHA512;
+    auto mgf1 = std::make_shared<MGF1>(mgf1DigestType);
```

Comment:
Maybe for the purpose of this test, we should get rid of all these details and just use the defaults.

Reply:
Done.

Reply:
Looks unchanged to me?